src/Microsoft.Diagnostics.Monitoring/ServerEndpointInfoSource.cs

```diff
@@ -0,0 +1,248 @@
+            CancellationToken linkedToken = linkedSource.Token;
+
+            // Prune connections that no longer have an active runtime instance before
+            // returning the list of connections.
+            await _endpointInfosSemaphore.WaitAsync(linkedToken).ConfigureAwait(false);
+            try
+            {
+                // Check the transport for each endpoint info and remove it if the check fails.
+                var endpointInfos = _endpointInfos.ToList();
+
+                var pruneTasks = new List<Task>();
+                foreach (IpcEndpointInfo info in endpointInfos)
+                {
+                    pruneTasks.Add(Task.Run(() => PruneIfNotViable(info, linkedToken), linkedToken));
+                }
+
+                await Task.WhenAll(pruneTasks).ConfigureAwait(false);
+
+                return _endpointInfos.Select(c => new EndpointInfo(c));
+            }
+            finally
+            {
+                _endpointInfosSemaphore.Release();
+            }
+        }
+
+        private async Task PruneIfNotViable(IpcEndpointInfo info, CancellationToken token)
+        {
+            using var timeoutSource = new CancellationTokenSource();
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, timeoutSource.Token);
+
+            try
+            {
+                timeoutSource.CancelAfter(PruneWaitForConnectionTimeout);
+
+                await info.Endpoint.WaitForConnectionAsync(linkedSource.Token).ConfigureAwait(false);
+            }
+            catch
+            {
+                // Only remove the endpoint info if due to some exception
+                // other than cancelling the pruning operation.
+                if (!token.IsCancellationRequested)
+                {
+                    _endpointInfos.Remove(info);
+                    OnRemovedEndpointInfo(info);
+                    _server.RemoveConnection(info.RuntimeInstanceCookie);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Accepts endpoint infos from the reversed diagnostics server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        private async Task ListenAsync(CancellationToken token)
+        {
+            // Continuously accept endpoint infos from the reversed diagnostics server so
+            // that <see cref="ReversedDiagnosticsServer.AcceptAsync(CancellationToken)"/>
+            // is always awaited in order to to handle new runtime instance connections
+            // as well as existing runtime instance reconnections.
+            while (!token.IsCancellationRequested)
+            {
+                try
+                {
+                    IpcEndpointInfo info = await _server.AcceptAsync(token).ConfigureAwait(false);
+
+                    _ = Task.Run(() => ResumeAndQueueEndpointInfo(info, token), token);
```

Comment:
You can create EventPipe sessions while the runtime is in the pause on start state.  By sending the Resume Runtime command for _every_ connection, you prevent this capability.  There should be some mechanism by which to either establish a session configuration _a priori_ or keep the runtime in the paused state until the user has sent any commands they want.  I'm fine with auto-resuming being the default to be clear.  I think we should make sure this is configurable in some way, though.

Reply:
These *EndpointInfoSource classes are meant as a means of collecting runtime instances and then performing some diagnostic operation on them far in the future. For example, dotnet-monitor will not interact with these instances until the user invokes the REST API to perform some diagnostic operation. So it makes sense to immediately resume their runtimes.

If it were to be configurable, how does a consumer of this class know whether or not to issue the ResumeRuntime command? It could just grab the endpoint, issue the command, and then perform the diagnostic operation _every single time_. Or maybe this class can expose another method for resuming the runtimes and recording that the runtime was already resumed (something like `void ResumeRuntime(Guid runtimeId)`) but that'll add additional state management to this class.

Reply:
Hmm, having thought about it, I think this is something that might make the most sense for `dotnet-monitor` right now.  Later, it could add this ability, but as long as the underlying code allows it, that's sufficient.